tests/test_id_generator.py

```diff
@@ -0,0 +1,142 @@
+        )
+        config = _TestConfig()
+        config.visit.packer.name = "observation"
+        config.visit.packer["observation"].n_observations = 10000
+        config.visit.packer["observation"].n_detectors = 99
+        config.visit.n_releases = 8
+        config.visit.release_id = 2
+        id_generator = config.visit.apply(data_id)
+        self.check_invariants(
+            id_generator,
+            unpacker=IdGenerator.unpacker_from_config(
+                config.visit, data_id.subset(self.universe.extract(["instrument"]))
+            ),
+            expected_release_id=2,
+        )
+
+    def test_exposure(self):
+        """Test an IdGenerator that packs {exposure, detector} data IDs."""
+        data_id = DataCoordinate.standardize(
+            instrument="I",
+            exposure=312,
+            detector=5,
+            universe=self.universe,
+        )
+        config = _TestConfig()
+        config.exposure.packer.name = "observation"
+        config.exposure.packer["observation"].n_observations = 10000
+        config.exposure.packer["observation"].n_detectors = 99
+        config.exposure.n_releases = 4
+        config.exposure.release_id = 3
+        id_generator = config.exposure.apply(data_id)
+        self.check_invariants(
+            id_generator,
+            unpacker=IdGenerator.unpacker_from_config(
+                config.exposure, data_id.subset(self.universe.extract(["instrument"]))
+            ),
+            expected_release_id=3,
+        )
+
+    def test_skymap(self):
+        """Test an IdGenerator that packs {tract, patch, band} data IDs."""
+        config = _TestConfig()
+        config.skymap.packer.n_tracts = 11
+        config.skymap.packer.n_patches = 9
+        data_id = DataCoordinate.standardize(
+            skymap="S", tract=9, patch=5, band="r", universe=self.universe
+        )
+        id_generator = config.skymap.apply(data_id)
+        self.check_invariants(
+            id_generator,
+            unpacker=IdGenerator.unpacker_from_config(
+                config.skymap, data_id.subset(self.universe.extract(["skymap"]))
+            ),
+        )
+
+    def check_invariants(
+        self,
+        id_gen: IdGenerator,
+        unpacker: Callable[[int], tuple[DataCoordinate, int]] | None = None,
+        expected_release_id: int = 0,
+    ):
+        """Check methods of the `IdGenerator` class for self-consistency and
+        expected values.
+        """
+        catalog = id_gen.make_source_catalog(self.schema)
+        for _ in range(5):
+            catalog.addNew()
+        array = id_gen.arange(1, 6)
+        self.assertEqual(list(catalog["id"]), list(array))
+        if unpacker is not None:
+            expected_data_id = id_gen.data_id
+            for i in range(5):
+                embedded_release_id, embedded_data_id, counter = unpacker(array[i])
+                self.assertEqual(counter, i + 1)
+                self.assertEqual(embedded_data_id, expected_data_id)
```

Comment:
I may be missing something, but I feel like this test is circular; it checks that the id you get out is what you passed in, no? Maybe that's all that is necessary and I'm missing something?

Reply:
It checks that the packing and unpacking methods are consistent with each other, and I figured that since they're effectively each independent implementations of the same algorithm, it wasn't worth it to write a third implementation in the test code.